src/chat/Chat.js

```diff
@@ -9,12 +9,28 @@ import MessageList from '../message-list/MessageList';
 import NoMessages from '../message/NoMessages';
 import ComposeBox from '../compose/ComposeBox';
 
-export default class MainScreen extends React.Component {
+export default class Chat extends React.Component {
+
+  unsentMessageIds = [];
+  lastSentTime = 0;
+
+  markAsRead = (messageIds) => {
+    const { auth, markMessagesRead, updateMessageFlags } = this.props;
+
+    this.unsentMessageIds.push(...messageIds);
+    markMessagesRead(this.unsentMessageIds);
```

Comment:
We should not mutate anything, though, not only in principle - which is important but also because this will not allow us to do the 'unread counts' later.

I did quite extensive benchmarking. It is much faster than you think... and slower than I thought. We need to fit in 16.7msecs to not drop a frame. For reference, cloning 1000 array elements takes like 0.02msecs. markMessagesRead func takes different times but it is usually in the 2msecs to 10msecs.

It looks good enough. If in practice we start seeing dropped frames, we can always add debouncing to the call itself, so it is called once the scrolling stops.